src/libxively/event_dispatcher/xi_time_event.c

```diff
@@ -0,0 +1,426 @@
+                                                         xi_vector_index_type_t lhs,
+                                                         xi_vector_index_type_t rhs,
+                                                         uint8_t elem_index_var_pos,
+                                                         int8_t dir )
+{
+    xi_vector_index_type_t* indexes[] = {&lhs, &rhs};
+
+    while ( lhs < rhs )
+    {
+        const xi_vector_index_type_t elem_index = *indexes[elem_index_var_pos];
+
+        const xi_time_event_t* lhs_element =
+            ( xi_time_event_t* )vector->array[dir < 0 ? elem_index : elem_index + dir]
+                .selector_t.ptr_value;
+        const xi_time_event_t* rhs_element =
+            ( xi_time_event_t* )vector->array[dir < 0 ? elem_index + dir : elem_index]
+                .selector_t.ptr_value;
+
+        if ( lhs_element->time_of_execution < rhs_element->time_of_execution )
+        {
+            /* if the elements are not in the right order lets swap them */
+            xi_swap_time_events( vector, lhs_element->position, rhs_element->position );
+
+            /* update the indexes */
+            *indexes[elem_index_var_pos] += dir;
+        }
+        else
+        {
+            /* thanks to the container invariant on the order of the elemements we can
+             * break at this moment */
+            break;
+        }
+    }
+
+    return *indexes[elem_index_var_pos];
+}
+
+/**
+ * @brief xi_time_event_bubble_and_sort_down
+ *
+ * Helper function used for insertion and re-insertion ( cancel opertaion ). It uses the
+ * bubble sort algorithm to put the element on the proper position.
+ *
+ * @note: invariant of this container - it's elements are sorted using the time of
+ * execution so:
+ * vector[i].time_of_execution <= vector[i+1].time_of_execution
+ *
+ * @param vector
+ * @param index
+ * @return new index of the element after bubbling it
+ */
+static xi_vector_index_type_t
+xi_time_event_bubble_and_sort_down( xi_vector_t* vector, xi_vector_index_type_t index )
+{
+    /* PRE-CONDITIONS */
+    assert( NULL != vector );
+    assert( index >= 0 );
+    assert( index < vector->elem_no );
+
+    return xi_time_event_bubble_core( vector, 0, index, 1, -1 );
+}
+
+/**
+ * @brief xi_time_event_bubble_and_sort_up
+ *
+ * Helper function used to restarting time event. It bubbles the time event element up to
+ * the end of the container. It works using the same principle as
+ * xi_time_event_bubble_and_sort_down function.
+ *
+ * @see xi_time_event_bubble_and_sort_down function xi_time_event_restart
+ *
+ * @param vector
+ * @param index
+ * @return
+ */
+static xi_vector_index_type_t
+xi_time_event_bubble_and_sort_up( xi_vector_t* vector, xi_vector_index_type_t index )
+{
+    /* PRE-CONDITIONS */
+    assert( NULL != vector );
+    assert( index >= 0 );
+    assert( index < vector->elem_no );
+
+    return xi_time_event_bubble_core( vector, index, vector->elem_no - 1, 0, 1 );
```

Comment:
yeah that's true, but this impression is true because those are values - if I used enums it would be more convenient but defining enums just for purpose of this internal helper ? It's not the API function.